integrations/parquet/src/async_reader.rs

```diff
@@ -0,0 +1,343 @@
+///     writer.write(&to_write).await.unwrap();
+///     writer.close().await.unwrap();
+///
+///     // gap: Allow the underlying reader to merge small IOs
+///     // when the gap between multiple IO ranges is less than the threshold.
+///     let reader = operator
+///         .reader_with(path)
+///         .gap(512 * 1024)
+///         .chunk(16 * 1024 * 1024)
+///         .concurrent(16)
+///         .await
+///         .unwrap();
+///     let content_len = operator.stat(path).await.unwrap().content_length();
+///     let reader = AsyncReader::new(reader, content_len).with_footer_size(512 * 1024);
+///     let mut stream = ParquetRecordBatchStreamBuilder::new(reader)
+///         .await
+///         .unwrap()
+///         .build()
+///         .unwrap();
+///     let read = stream.next().await.unwrap().unwrap();
+///     assert_eq!(to_write, read);
+/// }
+/// ```
+pub struct AsyncReader {
+    inner: Reader,
+    content_len: u64,
+    // The prefetch size for fetching file footer.
+    footer_size: usize,
+}
+
+fn set_footer_size(footer_size: usize, content_size: u64) -> usize {
+    let footer_size = max(footer_size, FOOTER_SIZE);
+    min(footer_size as u64, content_size) as usize
+}
+
+impl AsyncReader {
+    /// Create a [`AsyncReader`] by given [`Reader`].
+    pub fn new(reader: Reader, content_len: u64) -> Self {
+        Self {
+            inner: reader,
+            content_len,
+            footer_size: set_footer_size(PREFETCH_FOOTER_SIZE, content_len),
+        }
+    }
+
+    /// Set prefetch size for fetching file footer.
+    pub fn with_footer_size(mut self, footer_size: usize) -> Self {
+        self.footer_size = set_footer_size(footer_size, self.content_len);
+        self
+    }
+}
+
+impl AsyncFileReader for AsyncReader {
+    fn get_bytes(&mut self, range: Range<usize>) -> BoxFuture<'_, ParquetResult<bytes::Bytes>> {
+        async move {
+            Ok(self
+                .inner
+                .read(range.start as u64..range.end as u64)
+                .await
+                .map_err(|err| ParquetError::External(Box::new(err)))?
+                .to_bytes())
+        }
+        .boxed()
+    }
+
+    fn get_byte_ranges(
+        &mut self,
+        ranges: Vec<Range<usize>>,
+    ) -> BoxFuture<'_, ParquetResult<Vec<bytes::Bytes>>> {
+        async move {
+            Ok(self
+                .inner
+                .fetch(
+                    ranges
+                        .into_iter()
+                        .map(|range| range.start as u64..range.end as u64)
+                        .collect(),
+                )
+                .await
+                .map_err(|err| ParquetError::External(Box::new(err)))?
+                .into_iter()
+                .map(|buf| buf.to_bytes())
+                .collect::<Vec<_>>())
+        }
+        .boxed()
+    }
+
+    fn get_metadata(&mut self) -> BoxFuture<'_, ParquetResult<std::sync::Arc<ParquetMetaData>>> {
```

Comment:
Reading the metadata might be the slowest part for retrieving Parquet files from S3. Perhaps we should consider benchmarking this in the following PRs.

Reply:
Yes, it could be. In our system, we cache the file metadata in memory. Let's introduce a benchmark in future PRs.